addons/survey/controllers/main.py

```diff
@@ -732,35 +727,130 @@ def _get_user_input_domain(self, survey, line_filter_domain, **post):
+
+        * Model = 'A' : Match a `survey.question.answer` record (simple_choice, multiple_choice, matrix)
+        * Model = 'L' : Match a `survey.user_input.line` record (char_box, text_box, numerical_box, date, datetime)
+        :rtype: (collections.defaultdict[int, list[int]], list[int])
+        """
+        answer_cols_to_rows_mapping = defaultdict(list)
+        answer_user_input_line_ids = []
+
+        for data in post.get('filters', '').split('|'):
+            if not data:
+                break
+            model, row_id, ans_id = data.split(',')
+            row_id, ans_id = int(row_id), int(ans_id)
+            if model == 'A':
+                answer_cols_to_rows_mapping[ans_id].append(row_id)
+            elif model == 'L' and not row_id:
+                answer_user_input_line_ids.append(ans_id)
+
+        return answer_cols_to_rows_mapping, answer_user_input_line_ids
+
+    @staticmethod
+    def _get_value_with_operator(record, row_id=0):
+        """ Get the answer's value and domain operator to find answers with the
+        same value (depends on answer's model and question type).
+
+        :param record: 'survey.question.answer' or 'survey.user_input.line' record
+        :rtype: (string, string)
+        """
+        if record.question_id.question_type in ('simple_choice', 'multiple_choice', 'matrix'):
+            return f'{f"{row_id.value} : " if row_id else ""}{record.value}', None
+        return record[f'value_{record.question_id.question_type}'], {
+            'text_box': 'ilike',
+            'char_box': '=ilike',
+            'date': '=',
+            'datetime': '=',
+            'numerical_box': '=',
+        }[record.question_id.question_type]
+
+    @staticmethod
+    def _prepare_matching_answer_domain(question_type, answer, row_id=0, operator=None, value=None):
+        if question_type == "matrix":
+            return [expression.AND([[('question_id', '=', answer.question_id.id)], [('matrix_row_id', '=', row_id)],
+                                   [('suggested_answer_id', '=', answer.id)]])]
+        elif question_type in ('multiple_choice', 'simple_choice'):
+            return [expression.AND([[('question_id', '=', answer.question_id.id)], [('suggested_answer_id', '=', answer.id)]])]
+        elif question_type in ('char_box', 'text_box', 'numerical_box', 'date', 'datetime'):
+            return [expression.AND([[('question_id', '=', answer.question_id.id)], [(f'value_{answer.answer_type}', operator, value)]])]
+        return []
+
+    @classmethod
+    def _prepare_search_filter(cls, model, record, row_id=0, value=None):
+        """ Format parameters used to render/remove this filter on the results page.
+        'record' should be of 'survey.question.answer' or 'survey.user_input.line' models.
+
+        :rtype: dict
+        """
+        if value is None:
+            value = cls._get_value_with_operator(record, row_id=record.browse(row_id))[0]
+        return {
```

Comment:
Hmm, this is only used twice, takes strangely named parameters like "model", which is not really a model, record, which is again a mix of answers and lines, ...

I think we should not make a method and just repeat the keys when those entries are added. Seems like it adds confusion for a bit of code lines saved.

Reply:
(Side-note: I see you use "model" for those A/L characters in other places, would rename to something more meaningful like "model_short_key")